src/cv_compute/examples/examplesScript.sml

```diff
@@ -0,0 +1,216 @@
+ * related values.
+ * ------------------------------------------------------------------------- *)
+
+Definition factc_def:
+  factc n =
+    cv_if (cv_lt n (Num 1))
+          (Num 1)
+          (cv_mul n (factc (cv_sub n (Num 1))))
+Termination
+  WF_REL_TAC `measure cv_size_alt`
+  \\ Cases \\ simp [cv_size_alt_def, CaseEq "bool"]
+End
+
+Definition fact_def:
+  fact n = if n < 1 then 1 else n * fact (n - 1)
+End
+
+Theorem factc_is_fact:
+  !n. factc (Num n) = Num (fact n)
+Proof
+  Induct \\ once_rewrite_tac [factc_def, fact_def] \\ simp []
+QED
+
+(* cv_compute takes a list of compute equations, and an input term. All function
+ * constants in the term must have a code equation. Here, factc_def is the only
+ * code equation:
+ *)
+
+Triviality factc_test =
+  time (cv_compute [factc_def]) ``factc (Num 123)``;
+
+(* Same computation using EVAL:
+ *)
+
+Triviality fact_test = time EVAL ``fact 123``;
+
+(* Using cv_compute as a tactic:
+ *)
+
+Triviality fact_eq_factc:
+  factc (Num 12) = Num 479001600
+Proof
+  CONV_TAC (TOP_DEPTH_CONV (UNCHANGED_CONV (cv_compute [factc_def])))
+  \\ REFL_TAC
+QED
+
+(* -------------------------------------------------------------------------
+ * Example 2: primality checking using trial division
+ *
+ * The following functions create a list of primes up to some bound. Primes
+ * are found using trial division. Lists are encoded as pairs terminated by
+ * the number 0.
+ * ------------------------------------------------------------------------- *)
+
+Definition isprimec_aux_def:
+  isprimec_aux dvs n =
+    cv_if (cv_lt (cv_mul dvs (Num 2)) n)
+          (cv_if (cv_mod n dvs)
+                 (isprimec_aux (cv_add dvs (Num 2)) n)
+                 (Num 0))
+          (Num 1)
+Termination
+  WF_REL_TAC `measure (\(dvs,n). cv_size_alt (cv_sub n dvs))`
+  \\ Induct \\ simp [PULL_EXISTS]
+  \\ Cases_on `dvs` \\ gs [cv_size_alt_def] \\ rw [] \\ gs []
+End
+
+Definition isprimec_def:
+  isprimec n =
+    cv_if (cv_eq n (Num 2))
+          (Num 1)
+          (cv_if (cv_mod n (Num 2))
+                 (isprimec_aux (Num 3) n)
+                 (Num 0))
+End
+
+Definition primes_uptoc_def:
+  primes_uptoc upto =
+    cv_if (cv_ispair upto)
+          (Num 0)
+          (cv_if (cv_lt (Num 1) upto)
+                 (cv_if (isprimec upto)
+                        (Pair upto (primes_uptoc (cv_sub upto (Num 1))))
+                        (primes_uptoc (cv_sub upto (Num 1))))
+                 (Num 0))
+Termination
+  WF_REL_TAC `measure cv_size_alt`
+  \\ conj_tac \\ Cases \\ simp [cv_size_alt_def]
+End
+
+(* Here are the corresponding HOL functions, and a proof they compute values
+ * that are related to the values computed by the corresponding :cv functions.
+ *)
+
+Definition isprime_aux_def:
+  isprime_aux dvs n =
+    if dvs * 2 < n then
+      if n SAFEMOD dvs <> 0 then
+        isprime_aux (dvs + 2) n
+      else F
+    else T
+Termination
+  WF_REL_TAC `measure (\(dvs,n). n - dvs)`
+End
+
```

Comment:
Would be nice if this could point back to the core theories’ definition of `prime`.

Reply:
Done